### ISSUE TITLE
Lint negation "followed" by method call

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust-clippy#3903]
Calling `-x.foo()` will first call `foo` and then negate. While this is correct as per the precedence rules, it can be surprising (see https://github.com/rust-lang/rust/issues/59388). We should suggest a cleanup either by writing `(-x).foo()` if that was intended, by writing `-(x.foo())` for by introducing an intermediate variable:

```rust
let x_foo = x.foo();
-x_foo
```

**Comment by felix91gr** [comment]
Yeah, this makes a lot of sense. The precedence rules are perfect for compilers, but people don't necessarily have them ingrained. And code is read much more times than it is written :+1:

**Comment by Centril** [comment]
In my mind `-x.foo()` is no different than `!x.foo()`. This is the normal precedence of unary operators which many will understand. So I think this should not be in the complexity category as that would declare `-x.foo()` to be mostly unidiomatic.

**Comment by oli-obk** [comment]
I agree. The original issue was on numeric literals, not on variables. On variables it seems pretty clear. Linting `-12.foo()` but not `-x.foo()` seems like a middle ground to me.

Note that literals other than integers or floats should not be linted. `-"text".some_method()` is not negating the string literal and I think that's obvious enough

**Comment by memoryruins** [comment]
Numeric literals currently emit a precedence lint:

```rust
trait Trait { fn method(&self) -> i32 { 42 } }
impl Trait for &str {}

fn main() {
    let a = -42_i32.abs();
    let b = -42_f32.abs();
    let c = -"".method();
    dbg!(a, b, c);
}
```

```
warning: unary minus has lower precedence than method call
 --> src/main.rs:5:13
  |
5 |     let a = -42_i32.abs();
  |             ^^^^^^^^^^^^^ help: consider adding parentheses to clarify your intent: `-(42_i32.abs())`
  |
  = note: `#[warn(clippy::precedence)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#precedence

warning: unary minus has lower precedence than method call
 --> src/main.rs:6:13
  |
6 |     let b = -42_f32.abs();
  |             ^^^^^^^^^^^^^ help: consider adding parentheses to clarify your intent: `-(42_f32.abs())`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#precedence
  ```

Is there anything left to do for this issue?
